fendermint/vm/interpreter/src/genesis.rs

```diff
@@ -0,0 +1,795 @@
+
+                let params = ConstructorParameters::new(ipc_params.gateway, genesis.validators)
+                    .context("failed to create gateway constructor")?;
+
+                let facets = deployer
+                    .facets(ipc::gateway::CONTRACT_NAME)
+                    .context("failed to collect gateway facets")?;
+
+                deployer.deploy_contract(state, ipc::gateway::CONTRACT_NAME, (facets, params))?
+            };
+
+            // IPC SubnetRegistry actor.
+            {
+                use ipc::registry::ConstructorParameters;
+
+                let mut facets = deployer
+                    .facets(ipc::registry::CONTRACT_NAME)
+                    .context("failed to collect registry facets")?;
+
+                let getter_facet = facets.remove(0);
+                let manager_facet = facets.remove(0);
+                let rewarder_facet = facets.remove(0);
+                let checkpointer_facet = facets.remove(0);
+                let pauser_facet = facets.remove(0);
+                let diamond_loupe_facet = facets.remove(0);
+                let diamond_cut_facet = facets.remove(0);
+                let ownership_facet = facets.remove(0);
+
+                debug_assert_eq!(facets.len(), 2, "SubnetRegistry has 2 facets of its own");
+
+                let params = ConstructorParameters {
+                    gateway: gateway_addr,
+                    getter_facet: getter_facet.facet_address,
+                    manager_facet: manager_facet.facet_address,
+                    rewarder_facet: rewarder_facet.facet_address,
+                    pauser_facet: pauser_facet.facet_address,
+                    checkpointer_facet: checkpointer_facet.facet_address,
+                    diamond_cut_facet: diamond_cut_facet.facet_address,
+                    diamond_loupe_facet: diamond_loupe_facet.facet_address,
+                    ownership_facet: ownership_facet.facet_address,
+                    subnet_getter_selectors: getter_facet.function_selectors,
+                    subnet_manager_selectors: manager_facet.function_selectors,
+                    subnet_rewarder_selectors: rewarder_facet.function_selectors,
+                    subnet_checkpointer_selectors: checkpointer_facet.function_selectors,
+                    subnet_pauser_selectors: pauser_facet.function_selectors,
+                    subnet_actor_diamond_cut_selectors: diamond_cut_facet.function_selectors,
+                    subnet_actor_diamond_loupe_selectors: diamond_loupe_facet.function_selectors,
+                    subnet_actor_ownership_selectors: ownership_facet.function_selectors,
+                    creation_privileges: 0,
+                };
+
+                deployer.deploy_contract(state, ipc::registry::CONTRACT_NAME, (facets, params))?;
+            };
+        }
+
+        Ok(out)
+    }
+}
+
+fn contract_src(name: &str) -> PathBuf {
+    PathBuf::from(format!("{name}.sol"))
+}
+
+struct ContractDeployer<'a, DB> {
```

Comment:
Would it be suitable to move it elsewhere? As it is not strictly related only to genesis.

Reply:
but currently only genesis is using it, maybe we can move it if others are using it.